cpp/arcticdb/pipeline/write_frame.cpp

```diff
@@ -389,4 +393,54 @@ std::vector<SliceAndKey> flatten_and_fix_rows(
     return output;
 }
 
+folly::Future<std::vector<StreamSink::RemoveKeyResultType>> remove_slice_and_keys(
+        std::vector<SliceAndKey>&& slices, StreamSink& sink
+) {
+    std::vector<VariantKey> keys;
+    std::transform(
+            std::make_move_iterator(std::begin(slices)),
+            std::make_move_iterator(std::end(slices)),
+            std::back_inserter(keys),
+            [](SliceAndKey&& key) { return std::move(key).key(); }
+    );
+    return sink.remove_keys(std::move(keys));
+};
+
+folly::SemiFuture<std::vector<SliceAndKey>> rollback_on_quota_exceeded(
+        std::vector<folly::Try<SliceAndKey>>&& vec, const std::shared_ptr<stream::StreamSink>& sink
+) {
+    std::vector<SliceAndKey> succeeded;
+    std::optional<folly::exception_wrapper> exception;
+    bool has_quota_limit_exceeded = false;
+    succeeded.reserve(vec.size());
+    for (auto& k : vec) {
+        if (k.hasException()) {
+            if (!exception.has_value()) {
+                exception = k.exception();
+            }
+
+            if (!has_quota_limit_exceeded && k.exception().template is_compatible_with<QuotaExceededException>()) {
+                has_quota_limit_exceeded = true;
+            }
+
+        } else if (k.hasValue()) {
+            succeeded.emplace_back(std::move(k.value()));
+        }
+    }
+
+    if (has_quota_limit_exceeded) {
+        return remove_slice_and_keys(std::move(succeeded), *sink).thenValue([exception](auto&&) {
```

Comment:
So this rollbacks only on quota limit. Do we care about other kind of exceptions should it be fixed in another PR?